### DISCUSSION TITLE
Which function to use to match tracks in beets database?

### DISCUSSION THREAD

**Discussion by arsaboo** [beetbox/beets#4422] [Q&A]
I am trying to create a command to import playlists. Several sources (e.g., Spotify, Deezer) provide APIs to obtain the tracks in a given playlist. I can extract the `title`, `album`, and `year` for each track in the playlist. 

For Spotify, we can use `trackid`. However, sometimes the same track can have multiple ids (depending on the number of versions) and we may not get a match with the `trackid`. Thus, my approach will be to match the identifier, if available. Otherwise, use manual text matching (based on album, title, and year information).  

Now, I am wondering what is the best way to match that information in the beets database to get the `item`. Basically, I will pass the title, album, and year information and I would like to obtain a match (if available). I am assuming there is a function already somewhere that does the matching and returns the `item`. 

I should add that sometimes there may not be an exact match. For example, the Spotify title could be `Ye Tune Kya Kiya` and the title in the database could be `Ye Tune Kya Kiya Hai`. So, the closest match above a certain threshold (distance?) should be returned.

**Comment by sampsyo** [comment]
Sure! You'll want to use `library.items()` to issue a query:
https://github.com/beetbox/beets/blob/50825dbfc8443869ff2f1cad406e310a597c1aad/beets/library.py#L1539

For the `query` argument, you'll want to pass an `AndQuery` of `MatchQuery`s describing the criteria you need. There's an example in the `subsonicplaylist` plugin, for instance:
https://github.com/beetbox/beets/blob/018396ccf74a35e56ab346ad390476a0e347307a/beetsplug/subsonicplaylist.py#L72-L74

**Comment by arsaboo** [comment]
Thanks @sampsyo. Does it do some sort of "fuzzy" matching?

**Reply by sampsyo** [reply]
No, that's for an exact match. You'll want to use other query classes to get different kinds of matching (exactly which depends on the kind of "fuzziness" you are interested in).

**Reply by arsaboo** [reply]
Which one would you recommend to get the closest match based on `album` and `title`? There may not be an exact match, which makes it a little difficult. 

Ideally, something that would return an item with `levenshtein_distance` less than a certain threshold.

**Reply by sampsyo** [reply]
We don't have a built-in query that minimizes edit distance (i.e., something that sorts by proximity). However, the `fuzzy` plugin provides a kind of query that returns matches whose distance is below a threshold; that might provide inspiration:
https://github.com/beetbox/beets/blob/master/beetsplug/fuzzy.py

**Comment by arsaboo** [comment]
@sampsyo So, I tried this function:
```python
    def _update_recently_played(self, lib):
        """Fetch the Plex track key."""
        tracks = self.music.search(
            filters={'track.lastViewedAt>>': '7d'}, libtype='track')
        with lib.transaction():
            for track in tracks:
                query = MatchQuery('plex_ratingkey', track.ratingKey)
                item = lib.items(query)
                self._log.debug('Obtained item: {}', item)
                self._fetch_plex_info(item, write=True, force=True)
```
This is to obtain the item with matching `plex_ratingkey`. However, I keep getting this error:
```
Traceback (most recent call last):
  File "/home/arsaboo/.local/bin/beet", line 8, in <module>
    sys.exit(main())
  File "/home/arsaboo/.local/lib/python3.10/site-packages/beets/ui/__init__.py", line 1304, in main
    _raw_main(args)
  File "/home/arsaboo/.local/lib/python3.10/site-packages/beets/ui/__init__.py", line 1291, in _raw_main
    subcommand.func(lib, suboptions, subargs)
  File "/home/arsaboo/.local/lib/python3.10/site-packages/beetsplug/plexsync.py", line 128, in func_sync_recent
    self._update_recently_played(lib)
  File "/home/arsaboo/.local/lib/python3.10/site-packages/beetsplug/plexsync.py", line 236, in _update_recently_played
    item = lib.items(query)
  File "/home/arsaboo/.local/lib/python3.10/site-packages/beets/library.py", line 1541, in items
    return self._fetch(Item, query, sort or self.get_default_item_sort())
  File "/home/arsaboo/.local/lib/python3.10/site-packages/beets/library.py", line 1519, in _fetch
    return super()._fetch(
  File "/home/arsaboo/.local/lib/python3.10/site-packages/beets/dbcore/db.py", line 1093, in _fetch
    rows = tx.query(sql, subvals)
  File "/home/arsaboo/.local/lib/python3.10/site-packages/beets/dbcore/db.py", line 858, in query
    cursor = self.db._connection().execute(statement, subvals)
sqlite3.OperationalError: no such column: plex_ratingkey
```
I can confirm that the `plex_ratingkey` column exists. 
```
Item flexible attributes:
  data_source
  plex_guid
  plex_lastratedat
  plex_lastviewedat
  plex_ratingkey
  plex_skipcount
  plex_updated
```

**Comment by arsaboo** [comment]
I also tried a simple `item = lib.get_item(track.ratingKey)`, but that returns `None`. Basically, I am trying to get the item with `plex_ratingkey` (in the beets database) = `track.ratingKey`. See [this function](https://github.com/arsaboo/beets/blob/Plex_playlist2/beetsplug/plexsync.py#L229-L244) for context.

**Comment by arsaboo** [comment]
Ok....looks like I was missing `fast=False` in that query.

**Reply by sampsyo** [reply]
Ah yes, this would do it!!
